File: static/jd-parser.js
export function parseJdBlocks(text) {
  if (!text) return [];
  const lines = text.split('\n');

  // Skip leading boilerplate: nav links, metadata, short labels before the real JD prose.
  // "Real content" = a line that reads like a sentence (≥ 80 chars and contains punctuation
  // or common prose words), a recognisable section header keyword, or an emoji-led headline.
  const HEADER_RE = /^(overview|about|description|what you|role|position|responsibilities|qualifications|requirements|compensation|benefits|about us|the opportunity|the role|job summary|summary|who we|hiring)/i;
  const EMOJI_RE = /^\p{Emoji_Presentation}/u;
  const isProse = (l) => l.length >= 80 || (l.length >= 50 && /[,.]/.test(l));
  // For LinkedIn pages, "Feed post" marks the start of the actual post — skip profile chrome
  // before it by only counting lines that appear after it as candidate starts.
  const feedPostIdx = lines.findIndex(l => l.trim() === 'Feed post');
  const searchFrom = feedPostIdx >= 0 ? feedPostIdx + 1 : 0;

  let start = searchFrom;
  for (let i = searchFrom; i < Math.min(lines.length, searchFrom + 30); i++) {
    const l = lines[i].trim();
    if (isProse(l) || HEADER_RE.test(l) || EMOJI_RE.test(l)) { start = i; break; }
  }

  const blocks = [];
  let current = null;

  const flush = () => { if (current) { blocks.push(current); current = null; } };

  for (const raw of lines.slice(start)) {
    const line = raw.trim();

    if (!line) { flush(); continue; }

    // Stop at the LinkedIn concatenated duplicate — a very long run-on line that starts
    // with "Feed post" and contains the whole post without whitespace breaks.
    if (/^Feed post\S/.test(line)) { flush(); break; }

    // Bullet / list item
    const bulletContent = line.match(/^[•\-*◦·▪▸►▷→✅✓✔]\s+(.+)/)?.[1]
      ?? line.match(/^\d+[.)]\s+(.+)/)?.[1];
    if (bulletContent) {
      if (!current || current.type !== 'list') { flush(); current = { type: 'list', items: [] }; }
      current.items.push(bulletContent);
      continue;
    }

    // Section heading: ALL-CAPS short line, or line ending in ':', or known header words, or emoji-led
    const isHeading =
      (line.length < 80 && line === line.toUpperCase() && /[A-Z]{3}/.test(line) && !/[,;()\d]/.test(line))
      || (line.endsWith(':') && line.length < 70 && !line.includes('.'))
      || (line.length < 70 && HEADER_RE.test(line))
      || (EMOJI_RE.test(line) && line.length < 80);

    if (isHeading) {
      flush();
      blocks.push({ type: 'heading', text: line.replace(/:$/, '') });
      continue;
    }

    // Horizontal rule
    if (/^[-=]{3,}$/.test(line)) { flush(); blocks.push({ type: 'hr' }); continue; }

    // Paragraph text — merge consecutive lines into one paragraph
    if (!current || current.type !== 'paragraph') {
      flush();
      current = { type: 'paragraph', text: line };
    } else {
      current.text += ' ' + line;
    }
  }
  flush();

  // Drop trailing <hr> with nothing after it (LinkedIn separator before the duplicate)
  while (blocks.length && blocks[blocks.length - 1].type === 'hr') blocks.pop();

  return blocks;
}

File: core/Util/JDParser.swift
// JDParser.swift — port of static/jd-parser.js
// Pure Foundation only — no SwiftUI/AppKit/SwiftData imports.
// swiftlint:disable line_length
import Foundation

// MARK: - Types

public enum JDBlock: Equatable {
    case heading(text: String)
    case paragraph(text: String)
    case list(items: [String])
    case horizontalRule
}

// MARK: - Parser

/// Parses a job description plain text into structured blocks.
///
/// Mirrors jd-parser.js `parseJdBlocks`.
public func parseJdBlocks(_ text: String?) -> [JDBlock] {
    guard let text, !text.isEmpty else { return [] }

    let lines = text.components(separatedBy: "\n")

    // Skip leading boilerplate: find first "real content" line.
    let headerRe = try? NSRegularExpression(
        pattern: #"^(overview|about|description|what you|role|position|responsibilities|qualifications|requirements|compensation|benefits|about us|the opportunity|the role|job summary|summary|who we|hiring)"#,
        options: .caseInsensitive
    )
    let emojiRe = try? NSRegularExpression(pattern: #"^\p{Emoji_Presentation}"#)

    let isProse: (String) -> Bool = { str in
        str.count >= 80 || (str.count >= 50 && (str.contains(",") || str.contains(".")))
    }

    // For LinkedIn pages, "Feed post" marks the start of actual post content.
    let feedPostIdx = lines.firstIndex(where: { $0.trimmingCharacters(in: .whitespaces) == "Feed post" }) ?? -1
    let searchFrom = feedPostIdx >= 0 ? feedPostIdx + 1 : 0

    var start = searchFrom
    let searchLimit = min(lines.count, searchFrom + 30)
    for idx in searchFrom ..< searchLimit {
        let line = lines[idx].trimmingCharacters(in: .whitespaces)
        let nsStr = line as NSString
        let range = NSRange(location: 0, length: nsStr.length)
        let matchesHeader = headerRe?.firstMatch(in: line, range: range) != nil
        let matchesEmoji = emojiRe?.firstMatch(in: line, range: range) != nil
        if isProse(line) || matchesHeader || matchesEmoji {
            start = idx
            break
        }
    }

    var blocks: [JDBlock] = []
    var current: JDBlock?

    let flush: () -> Void = {
        if let cur = current { blocks.append(cur); current = nil }
    }

    for raw in lines[start...] {
        let line = raw.trimmingCharacters(in: .whitespaces)

        if line.isEmpty {
            flush()
            continue
        }

        // Stop at LinkedIn concatenated duplicate — long run-on starting with "Feed post" + non-space
        if line.hasPrefix("Feed post") && line.count > "Feed post".count && !line[line.index(
            line.startIndex,
            offsetBy: "Feed post".count
        )].isWhitespace {
            flush()
            break
        }

        // Bullet / list item
        if let bulletContent = extractBulletContent(line) {
            if case let .list(items) = current {
                current = .list(items: items + [bulletContent])
            } else {
                flush()
                current = .list(items: [bulletContent])
            }
            continue
        }

        // Horizontal rule
        if isHorizontalRule(line) {
            flush()
            blocks.append(.horizontalRule)
            continue
        }

        // Section heading
        let nsLine = line as NSString
        let lineRange = NSRange(location: 0, length: nsLine.length)
        let matchesHeader = headerRe?.firstMatch(in: line, range: lineRange) != nil
        let matchesEmoji = emojiRe?.firstMatch(in: line, range: lineRange) != nil
        let isHeading = isHeadingLine(line, matchesHeader: matchesHeader, matchesEmoji: matchesEmoji)

        if isHeading {
            flush()
            let headingText = line.hasSuffix(":") ? String(line.dropLast()) : line
            blocks.append(.heading(text: headingText))
            continue
        }

        // Paragraph — merge consecutive lines
        if case let .paragraph(existing) = current {
            current = .paragraph(text: existing + " " + line)
        } else {
            flush()
            current = .paragraph(text: line)
        }
    }
    flush()

    // Drop trailing <hr> blocks (LinkedIn separator before duplicate)
    while case .horizontalRule = blocks.last {
        blocks.removeLast()
    }

    return blocks
}

// MARK: - Helpers

private func extractBulletContent(_ line: String) -> String? {
    // Unicode bullet characters: •-*◦·▪▸►▷→✅✓✔
    let bulletChars: Set<Character> = ["•", "-", "*", "◦", "·", "▪", "▸", "►", "▷", "→", "✅", "✓", "✔"]
    if let first = line.first, bulletChars.contains(first) {
        let rest = line.dropFirst().trimmingCharacters(in: .whitespaces)
        if !rest.isEmpty { return rest }
    }
    // Numbered: \d+[.)]\s+...
    if let match = line.range(of: #"^\d+[.)]\s+"#, options: .regularExpression) {
        let rest = String(line[match.upperBound...])
        if !rest.isEmpty { return rest }
    }
    return nil
}

private func isHorizontalRule(_ line: String) -> Bool {
    guard line.count >= 3 else { return false }
    return line.allSatisfy { $0 == "-" } || line.allSatisfy { $0 == "=" }
}

private func isHeadingLine(_ line: String, matchesHeader: Bool, matchesEmoji: Bool) -> Bool {
    let isAllCaps = line.count < 80
        && line == line.uppercased()
        && line.contains(where: \.isUppercase)
        && line.contains(where: { $0.isLetter && $0.isUppercase })
        // Must have at least 3 uppercase letters
        && line.count(where: { $0.isUppercase }) >= 3
        && !line.contains(",") && !line.contains(";") && !line.contains("(")
        && !line.contains(")") && !line.contains(where: \.isNumber)

    let endsWithColon = line.hasSuffix(":")
        && line.count < 70
        && !line.contains(".")

    let isKnownHeader = line.count < 70 && matchesHeader
    let isEmojiLed = matchesEmoji && line.count < 80

    return isAllCaps || endsWithColon || isKnownHeader || isEmojiLed
}

// swiftlint:enable line_length cyclomatic_complexity function_body_length
